### PR TITLE
UTF-8: Add partial support for parsing UTF-8 metric and label names

### PR DESCRIPTION
**Pull request by ywwg**
This PR adds support for the new grammar of `{"metric_name", "l1"="val"}` to promql and some of the exposition formats. This grammar will also be valid for non-UTF-8 names. UTF-8 names will not be considered valid unless model.NameValidationScheme is changed.

This PR does not update the go expfmt parser in text_parse.go, which will be addressed by https://github.com/prometheus/common/issues/554/.

Correction, this PR *does* update the promql parser.

Part of https://github.com/prometheus/prometheus/issues/13095

### CODE REVIEW

File: scrape/scrape_test.go [hunk 1]
```diff
@@ -1331,7 +1331,7 @@ func TestScrapeLoopAppend(t *testing.T) {
 			// Honor Labels should ignore labels with the same name.
 			title:           "Honor Labels",
 			honorLabels:     true,
-			scrapeLabels:    `metric{n1="1" n2="2"} 0`,
+			scrapeLabels:    `metric{n1="1", n2="2"} 0`,
```
Comment (beorn7) [inline scrape/scrape_test.go:1334]: Does the old way still work? As discussed, we should not change this behavior without a deeper discussion.

**Review by beorn7 — CHANGES_REQUESTED** [review]
Generally looks good. Mostly nitpicking in the comment and requesting more test cases.

I don't really feel I'm the right person to review this because I am not familiar at all with the generated parser. I might also ask dumb questions in my comments.

Let's try to find a reviewer that knows those things better.

**Comment by beorn7** [comment]
I believe the Go tests should pass if you run `go mod tidy`.

**Comment by beorn7** [comment]
> This PR does not update PromQL parsing

It does now, doesn't it?

However, the frontend cod isn't updated yet, so while things work, the UI is complaining passive-aggressively with squiggly lines:

image

**Comment by ywwg** [comment]
> It does now, doesn't it?

fixed

File: scrape/scrape_test.go (hunk 1 above)
Comment (ywwg) [reply scrape/scrape_test.go:1334]: Yes, the old way still does work... I guess I will leave in one example of this test, but I want to add a note that the syntax is invalid and may be removed if we decide it's a problem.

File: promql/parser/generated_parser.y [hunk 2]
```diff
@@ -582,7 +582,11 @@ label_match_list: label_match_list COMMA label_matcher
                 ;
 
 label_matcher   : IDENTIFIER match_op STRING
-                        { $$ = yylex.(*parser).newLabelMatcher($1, $2, $3);  }
+                        { $$ = yylex.(*parser).newLabelMatcher($1, $2, $3); }
+                | string_identifier match_op STRING
```
Comment (darshanime) [inline promql/parser/generated_parser.y:586]: do we need `string_identifier match_op error` rule? or is the parser error useful without it already?

File: model/textparse/openmetricsparse_test.go [hunk 3]
```diff
@@ -300,6 +300,131 @@ foo_total 17.0 1520879607.789 # {id="counter-test"} 5`
...
+	input := `# HELP "go.gc_duration_seconds" A summary of the GC invocation durations.
+# TYPE "go.gc_duration_seconds" summary
+# UNIT "go.gc_duration_seconds" seconds
+{"go.gc_duration_seconds",quantile="0"} 4.9351e-05
+{"go.gc_duration_seconds",quantile="0.25"} 7.424100000000001e-05
+{"go.gc_duration_seconds",quantile="0.5",a="b"} 8.3835e-05
+{"http.status",q="0.9",a="b"} 8.3835e-05
+{"http.status",q="0.9",a="b"} 8.3835e-05
+{q="0.9","http.status",a="b"} 8.3835e-05
+{"go.gc_duration_seconds_sum"} 0.004304266`
```
[the remark this answers is not in the document]
Comment (ywwg) [reply model/textparse/openmetricsparse_test.go:319]: oo, found some bugs, thanks :)

File: model/textparse/promparse.go [hunk 4]
```diff
@@ -433,6 +502,14 @@ func yoloString(b []byte) string {
 	return *((*string)(unsafe.Pointer(&b)))
 }
 
+func (p *PromParser) verifyMetricName() error {
```
[the remark this answers is not in the document]
Comment (ywwg) [reply model/textparse/promparse.go:505]: Yes you're right, removed

File: promql/parser/generated_parser.y (hunk 2 above)
Comment (ywwg) [reply promql/parser/generated_parser.y:586]: would that be like
`{"x"=}` ?

Comment (ywwg) [reply promql/parser/generated_parser.y:586]: ok yes, adding that makes the error a little better

**Comment by ywwg** [comment]
test failure is I think unrelated but I'll dig in

**Comment by ywwg** [comment]
Hm this test failure is a bad sign -- there have been multiple places where people are testing the value of the negotiated format against the string constants.  This is Bad because there are now many different ways the formats can look and be equivalent. How do we prevent people from doing these bad string comparisons? make the string constants non-exported?

**Comment by ywwg** [comment]
fuzz error looks concerning, I'm trying to figure out how to run that test myself

**Comment by ywwg** [comment]
I was able to debug the fuzz error, fixing

**Comment by ywwg** [comment]
got it

File: go.mod [hunk 5]
```diff
@@ -52,7 +52,7 @@ require (
 	github.com/prometheus/alertmanager v0.26.0
 	github.com/prometheus/client_golang v1.18.0
 	github.com/prometheus/client_model v0.5.0
-	github.com/prometheus/common v0.46.0
+	github.com/prometheus/common v0.46.1-0.20240206181200-773d5664eb8d
```
Comment (beorn7) [inline go.mod:55]: Let's not pin to specific commits here. I have just tagged 0.47.0 in common, so you can use it now.

File: model/textparse/openmetricsparse.go [hunk 6]
```diff
@@ -476,3 +519,11 @@ func (p *OpenMetricsParser) getFloatValue(t token, after string) (float64, error
 	}
 	return val, nil
 }
+
+func (p *OpenMetricsParser) verifyMetricName(start, end int) error {
```
[the remark this answers is not in the document]
Comment (beorn7) [reply model/textparse/openmetricsparse.go:535]: Apparently you removed `verifyMetricName` from promparse.go, but left it in here. But this should go, too, shouldn't it?

**Review by beorn7 — CHANGES_REQUESTED** [review]
Looks good now with fixed fuzzing and extended tests, as far as I can say.

See comments for minor things, otherwise I'd say this is good to go. Last chance for others to chime in!

File: go.mod (hunk 5 above)
Comment (ywwg) [reply go.mod:55]: I don't see the tag on github, do you need to push it?

**Comment by ywwg** [comment]
`Apparently you removed verifyMetricName from promparse.go, but left it in here. But this should go, too, shouldn't it?`

ah yes sorry

File: go.mod (hunk 5 above)
Comment (beorn7) [reply go.mod:55]: I have pushed it an hour ago, and even created a nice release.

Maybe Github is once more very eventually consistent?
https://github.com/prometheus/common/releases/tag/0.47.0

**Comment by beorn7** [comment]
Thank you. I think you can do the squashing and DCO'ing now, and then we can merge (barring any new objections from others).

File: promql/promql_test.go [hunk 7]
```diff
@@ -38,9 +38,9 @@ func newTestEngine() *Engine {
 	})
 }
 
-func TestEvaluations(t *testing.T) {
-	RunBuiltinTests(t, newTestEngine())
-}
+// func TestEvaluations(t *testing.T) {
```
Comment (ywwg) [inline promql/promql_test.go:41]: oops

**Review by beorn7 — APPROVED** [review]

**Comment by beorn7** [comment]
:tada:

**Pull request merged**